modules/service/src/main/resources/db/migration/V0928__gmos_binning.sql

```diff
@@ -0,0 +1,224 @@
+  END CASE;
+
+  RETURN result;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+--
+-- The effective width of the object, which is the object size estimate but
+-- limited by the slit width.
+--
+CREATE OR REPLACE FUNCTION calculate_effective_width(
+  iq          numeric(2, 1),
+  slit_µas    d_angle_µas,
+  src_profile jsonb
+)
+RETURNS d_angle_µas AS $$
+DECLARE
+  obj_size_µas d_angle_µas;
+BEGIN
+  SELECT calculate_object_size(iq, src_profile) INTO obj_size_µas;
+  RETURN least(obj_size_µas, slit_µas);
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+--
+-- GMOS longslit spectral binning calculation.
+--
+CREATE OR REPLACE FUNCTION calculate_gmos_long_slit_spectral_binning(
+  iq            numeric(2, 1),
+  slit_µas      d_angle_µas,
+  dispersion_pm smallint,
+  resolution    smallint,
+  blaze_pm      smallint,
+  src_profile   jsonb,
+  sampling      double precision
+)
+RETURNS smallint AS $$
+DECLARE
+  eff_width_µas d_angle_µas;
+  eff_res_as    double precision;
+  n_pix         double precision;
+  binning       smallint[];
+  bin           smallint;
+BEGIN
+  SELECT calculate_effective_width(iq, slit_µas, src_profile) INTO eff_width_µas;
+  eff_res_as := resolution::double precision / 2.0 / (eff_width_µas::double precision / 1000000.0);
+  n_pix      := (blaze_pm::double precision / 1000.0) / eff_res_as / (dispersion_pm / 1000.0);
+
+  SELECT ARRAY(SELECT c_count FROM t_gmos_binning ORDER BY c_count DESC) INTO binning;
+  FOREACH bin IN ARRAY binning
+  LOOP
+    IF sampling < (n_pix / bin::double precision) THEN
+      RETURN bin;
+    END IF;
+  END LOOP;
+
+  RETURN 1;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+--
+-- GMOS longslit spatial binning calculation.
+--
+CREATE OR REPLACE FUNCTION calculate_gmos_long_slit_spatial_binning(
+  iq              numeric(2, 1),
+  pixel_scale_µas d_angle_µas,
+  src_profile     jsonb,
+  sampling        double precision
+)
+RETURNS smallint AS $$
+DECLARE
+  obj_size_µas  d_angle_µas;
+  n_pix         double precision;
+  binning       smallint[];
+  bin           smallint;
+BEGIN
+  SELECT calculate_object_size(iq, src_profile) INTO obj_size_µas;
+  n_pix := (obj_size_µas::double precision / 1000000.0) /
+           (pixel_scale_µas::double precision / 1000000.0 * sampling);
+
+  SELECT ARRAY(SELECT c_count FROM t_gmos_binning ORDER BY c_count DESC) INTO binning;
+  FOREACH bin IN ARRAY binning
+  LOOP
+    IF bin::double precision < n_pix THEN
+      RETURN bin;
+    END IF;
+  END LOOP;
+
+  RETURN 1;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION lookup_bin_tag(
+  bin smallint
+)
+RETURNS d_tag AS $$
+BEGIN
+  RETURN (SELECT c_tag FROM t_gmos_binning WHERE c_count = bin);
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+--
+-- (xbin, ybin) tuple
+--
+CREATE TYPE gmos_binning AS (
+  xbin d_tag,
+  ybin d_tag
+);
+
+-- Set the default binning.
+CREATE OR REPLACE PROCEDURE set_gmos_long_slit_default_binning(
+  oid d_observation_id
+) AS $$
```

Comment:
We need to calculate the default binning in order to correct group observing modes.  Usually the binning is not explicitly set but we need to know what binning would be use.  As you can see, this calculation is rather involved. It makes me nervous to put so much logic into `plpgsql`.  Other options that occur to me:

* Carefully update all the mutations that touch relevant fields and explicitly redo the calculation and update via Scala code.
* Create a channel for updates to anything that would impact the default binning, respond to updates in Scala code.

What would be best?